### PR TITLE
[WIP] Esy-enabled Windows Build

### PR DESCRIPTION
**Pull request by bryphe**
Hi @aantron ,

Following from #43 - here's what I've been trying to get `luv` building on Windows with `esy`.

Unfortunately it's not merge-able at the moment - need to figure out how to preserve the existing OPAM build with `esy` on the side. I had some issues getting `uv` to build in the Cygwin + MingW sandbox - I ended up creating this project: <https://github.com/revery-ui/esy-libuv> to help build it. I had problems with both the `CMake` build in that environment, and with `gyp` - so I fell back to `automake`.

But hopefully we can use this as a starting point to figure out how to merge and reconcile the strategies, and give you an idea of some of the issues I faced on Windows.

It actually builds cross-platform:
image

I'm not an expert on `dune`, so I had to fall back on a `discover.ml` to pick out some of the build flags. But there is probably a nicer way to do it 😄 

And some of the examples work on Windows. Next step is to investigate the tests and see what the state is on all platforms.

Thanks again for all your work on these bindings! Really great to have this in the OCaml / Reason ecosystem.

cc @glennsl @ulrikstrid @lessp who were also interested in the Windows status

### CODE REVIEW

File: src/c/generate_c_functions.ml [hunk 1]
```diff
@@ -4,6 +4,14 @@
...
 
 let () =
+  
+  (* Include headers necessary for Windows *)
+  if Sys.win32 then
+    (* Setting a WINVER to Vista+ is necessary to get types like AI_ADDRCONFIG *)
+    (print_endline "#define WINVER 0x0600";
+    print_endline "#define _WIN32_WINNT 0x0600";
+    print_endline "#include <ws2tcpip.h>");
+
```
Comment (bryphe) [inline src/c/generate_c_functions.ml]: There were some headers we need to pull in to get some APIs (like `AI_ADDRCONFIG`). They were introduced in Vista, so we have to set the `WINVER` and `_WIN32_WINNT` values to `0x0600` (Vista)

File: src/c/luv_c_type_descriptions.ml [hunk 2]
```diff
@@ -126,7 +126,9 @@ struct
     module Option =
     struct
       let block_signal = constant "UV_LOOP_BLOCK_SIGNAL" int
-      let sigprof = constant "SIGPROF" int
+      let sigprof = match Sys.win32 with 
+      | false -> constant "SIGPROF" int
```
Comment (bryphe) [inline src/c/luv_c_type_descriptions.ml]: I'm not sure the best way to handle this, but `SIGPROF` isn't defined on Windows.

File: src/file.ml [hunk 3]
```diff
@@ -121,9 +121,11 @@ struct
     | `IWOTH
     | `IXOTH
 
+(* TODO: Not supported on Windows
     | `ISUID
     | `ISGID
     | `ISVTX
+*)
```
Comment (bryphe) [inline src/file.ml]: Likewise - several of these file constants were not defined on Windows.

**Comment by aantron** [comment]
Thanks. Going to be looking into this in more and more detail in the coming days.

Concerning the build system, it looks like https://github.com/libuv/libuv/pull/2682 will remove the libuv gyp build in the coming releases anyway. So, Luv will definitely be using another method of building libuv.

**Comment by aantron** [comment]
@bryphe Can you give any detail about the environment in which the Windows build is running? I'm starting out with a basic port using opam-repository-mingw in Cygwin, and opam, but I'd like to get your exact commands and any relevant bits of the setup, so I can make sure that works, too.

File: .ci/esy-build-steps.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+steps:
+  - task: NodeTool@0
+    inputs:
+      versionSpec: '8.9'
+  - script: npm install -g esy@0.6.2
+    displayName: 'npm install -g esy@0.6.2'
+  - script: esy install
+    displayName: 'esy install'
+  - script: esy build
+    displayName: 'esy build'
```
Comment (aantron) [inline .ci/esy-build-steps.yml]: Is it correct that these steps are running in `cmd`?

Comment (bryphe) [reply .ci/esy-build-steps.yml]: It is correct that they are running in `cmd` - one of the main benefits of using `esy`, for me, is that I don't have to change how I work as a windows developer. 

The set of steps described in `esy-build-steps.yml` is the same for OSX, Linux, and Windows - it's really nice to have a consistent workflow across all platforms:
- Install `esy` (`npm install -g esy`)
- Install dependencies (`esy install`)
- Build (`esy build`)

**Comment by bryphe** [comment]
> @bryphe Can you give any detail about the environment in which the Windows build is running?

@aantron - definitely! Thank you for your help with this.

`esy`, on Windows, under the hood provisions a Cygwin + Mingw64 environment very similar to the instructions here: https://fdopen.github.io/opam-repository-mingw/installation/

So even though esy itself can be run via `cmd`, `powershell`, etc - it delegates the actual build commands to the Cygwin+Mingw64 environment to build binaries. There are some POSIX utilities to support OCaml development (https://github.com/bryphe/esy-bash/blob/e93b90541dbf754bdbc64e249ab6580069f4492c/build-cygwin.js#L27) - essentially the bare minimum to build the OCaml compiler and support `esy`. Things like `bash`, `patch`, etc are available.

**Comment by bryphe** [comment]
Oh and I almost forgot, some more in-depth details here: https://github.com/esy/esy/pull/487/files

**Comment by aantron** [comment]
Progress update: I'm currently looking into how portable a build we can get using autotools, since this would be the most portable, if `autogen.sh` can be removed from the build steps for release mode. Ideally, we would run only a `./configure` and tell it to call the C compiler through the `ocaml` command, which would automatically make libuv use the right toolchain. This will still depend on `bash` and a few commands, but esy is likely to have everything installed in its internal Cygwin environment.

I also want to mention, the gyp build might have partially or fully succeeded with a patch to the esy distribution. It seems esy is distributing a Cygwin with only Python 2.7 installed:

```
$ EsyBash.exe 'ls /bin/python*'
/bin/python  /bin/python2  /bin/python2.7.exe
```

However, since both Luv and libuv are moving away from gyp, I'm not prepared to lobby for esy distributing Python 3, or for trying Python 2 with gyp again. This may be relevant for some other project in the future, though, so I wanted to note it.

**Comment by bryphe** [comment]
> Progress update: I'm currently looking into how portable a build we can get using autotools, 

Awesome! Thanks @aantron for looking at this. 

FYI - I can verify that the autotools approach works with the MingW toolchain - we use it for the windows build script: https://github.com/revery-ui/esy-libuv/blob/89d293b414c058f423dfe63ad80858b8d82cddb9/esy/build-windows.sh#L5

> Ideally, we would run only a ./configure and tell it to call the C compiler through the ocaml command, which would automatically make libuv use the right toolchain. 

That's a clever approach to get the right C compiler 👍 

> However, since both Luv and libuv are moving away from gyp, I'm not prepared to lobby for esy distributing Python 3, or for trying Python 2 with gyp again. 

Ya... the python versioning is always a pain. We encountered similar issues getting skia to build (which also requires `gyp`) - we used a version of `gyp` that required python2, and had to check and validate that: https://github.com/revery-ui/esy-skia/blob/master/esy/build.sh#L6 Not that great...

cc @prometheansacrifice - thought you might be interested in this work too

**Comment by aantron** [comment]
> I can verify that the autotools approach works with the MingW toolchain

Thanks! I'll put the build system patch I came up with soon-ish :) I'll probably do a release of luv at that point.

> we used a version of gyp that required python2

That's what I originally had in this repo, too, but at release time, I found out that some popular Linux distros no longer distribute Python 2 by default in their latest releases anymore, and that Python 2 end of support is this year (IIRC). So, I upgraded to gyp to the Python 3 version.

**Comment by prometheansacrifice** [comment]
Indeed I am! Infact the first time I cloned luv and built was on Windows. Ever since, I have been trying to take out a weekend for this, but great to see that work has started already! I'll try to look into this again hopefully this weekend

**Comment by aantron** [comment]
Closed in favor of https://github.com/aantron/luv/compare/9704de4aed85941a5769d59715a5219fd7485ec8...0c23e803e484eaf4d1bc1fcb6f250fa96fad263d. I used this PR for reference in places, thanks!

The Windows build now works, most tests pass, linking against Luv (and its vendored libuv) works. There is Windows CI for both [opam](https://github.com/aantron/luv/runs/507303503?check_suite_focus=true) and [esy](https://github.com/aantron/luv/runs/507303519?check_suite_focus=true) in a new GitHub actions setup for this repo.

I'll do the libuv 1.35.0 upgrade (#46), and then a Luv release.

With the way I did the port, there should be no need for a separate esy-libuv repo, unless, of course, you'd like to have libuv for a reason other than wrapping it in Luv. The `luv` package builds everything it needs internally on both opam and esy.

File: src/c/generate_c_functions.ml (hunk 1 above)
Comment (SatoAoaka) [reply src/c/generate_c_functions.ml:14]: ✅ 修正しました！

コミット cf09c4e で対応しました。
`server_default=text("NO_EVALUATION")` に変更し、シングルクォートを削除しました。

ご指摘ありがとうございました。

**Pull request closed without merging**